device/core/src/sas_authentication_provider.ts

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+import { EventEmitter } from 'events';
+import { TokenAuthenticationProvider, AuthenticationType, SharedAccessSignature, DeviceCredentials } from 'azure-iot-common';
+
+export class SharedAccessSignatureAuthenticationProvider extends EventEmitter implements TokenAuthenticationProvider {
+  type: AuthenticationType = AuthenticationType.Token;
+  automaticRenewal: boolean = false;
+  private _credentials: DeviceCredentials;
+
+  /**
+   * @private
+   *
+   * Initializes a new instance of the TokenAuthenticationProvider - users should only use the factory methods though.
+   *
+   * @param credentials       Credentials to be used by the device to connect to the IoT hub.
+   * @param securityProvider  Object used to sign the tokens that are going to be used during authentication.
+   */
+  constructor(credentials: DeviceCredentials) {
+    super();
+    /*Codes_SRS_NODE_SAS_AUTHENTICATION_PROVIDER_16_001: [The `constructor` shall store the credentials passed in the `credentials` argument.]*/
+    this._credentials  = credentials;
+  }
+
+  getDeviceCredentials(callback: (err: Error, credentials: DeviceCredentials) => void): void {
+    /*Codes_SRS_NODE_SAS_AUTHENTICATION_PROVIDER_16_002: [The `getDeviceCredentials` method shall call its callback with a `null` error parameter and the stored `credentials` object containing the current device credentials.]*/
+    callback(null, this._credentials);
+  }
+
+  updateSharedAccessSignature(sharedAccessSignature: string): void {
+    /*Codes_SRS_NODE_SAS_AUTHENTICATION_PROVIDER_16_003: [The `updateSharedAccessSignature` method shall update the stored credentials with the new `sharedAccessSignature` value passed as an argument.]*/
+    this._credentials.sharedAccessSignature = sharedAccessSignature;
+    /*Codes_SRS_NODE_SAS_AUTHENTICATION_PROVIDER_16_004: [The `updateSharedAccessSignature` method shall emit a `newTokenAvailable` event with no arguments.]*/
+    this.emit('newTokenAvailable');
+  }
+
+  static fromSharedAccessSignature(sharedAccessSignature: string): SharedAccessSignatureAuthenticationProvider {
+    if (!sharedAccessSignature) {
+      /*Codes_SRS_NODE_SAS_AUTHENTICATION_PROVIDER_16_005: [The `fromSharedAccessSignature` method shall throw a `ReferenceError` if the `sharedAccessSignature` argument is falsy.]*/
+      throw new ReferenceError('sharedAccessSignature cannot be \'' + sharedAccessSignature + '\'');
+    }
+    const sas = SharedAccessSignature.parse(sharedAccessSignature);
+    const decodedUri = decodeURIComponent(sas.sr);
+    const uriSegments = decodedUri.split('/');
```

Comment:
>decodedUri [](start = 24, length = 10)

We're not doing any validation on the format of the URI, and we might not be able to.  This isn't the worst thing in the world, but we're relying on a failure somewhere else to catch this, probably with an unrelated error.  Would it be better to fail here with a meaningful error if we could?  At the very least, we could maybe fail based on the number of segments.

Reply:
that is something we can revisit later.

---
In reply to: [157024476](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157024476) [](ancestors = 157024476)